src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs

```diff
@@ -19,9 +19,9 @@ public static partial class JsonSerializer
         /// <typeparamref name="TValue"/> is not compatible with the JSON,
         /// or when there is remaining data in the Stream.
         /// </exception>
-        public static TValue Deserialize<TValue>(ReadOnlySpan<byte> utf8Json, JsonSerializerOptions options = null)
+        public static TValue Deserialize<TValue>(ReadOnlySpan<byte> utf8Json, JsonSerializerOptions? options = null)
         {
-            return (TValue)ParseCore(utf8Json, typeof(TValue), options);
+            return (TValue)ParseCore(utf8Json, typeof(TValue), options)!;
```

Comment:
We should probably do some null check before the cast to avoid nullref for value types or non-nullable reference types.

```C#
public struct Foo
{
}

Foo mystruct = JsonSerializer.Deserialize<Foo>("null"); // throws NRE rather than InvalidOperationException
```

Reply:
> We should probably do some null check before the cast to avoid nullref for value types or non-nullable reference types.

All reference types are nullable at runtime.  (Non-nullable reference types are just a compiler assistance feature).  So this probably needs `[MaybeNull]` on the return.

Reply:
> We should probably do some null check before the cast to avoid nullref for **value types** or non-nullable reference types.

Are you sure  for **value types** it can return null.

> So this probably needs [MaybeNull] on the return.

That was the question @stephentoub asked about this method https://github.com/dotnet/runtime/pull/114#discussion_r348872541. If this can return null for non null ref types. For me by following the call stack I didn't see a case returning default value, but it is branching a lot, very difficult to follow all branching.

Reply:
> Are you sure for value types it can return null.

Yes, in my example above `Foo` is a struct type. That's why we end up with a null reference exception.

> If this can return null for non null ref types. 

It can. However, @safern suggested earlier not annotating the `Deserialize` with `MaybeNull`.
https://github.com/dotnet/runtime/pull/528#discussion_r357484307

```C#
#nullable enable 
using System;
using System.Text.Json;

namespace NullabilityCheck
{
    class Program
    {
        static void Main(string[] args)
        {
            string str = JsonSerializer.Deserialize<string>("null");
            Console.WriteLine(str == null); // Outputs: true
        }
    }
}
```

Reply:
> We should probably do some null check before the cast to avoid nullref for value types or non-nullable reference types.

https://github.com/dotnet/runtime/issues/1037

Reply:
Yes NRE should never be thrown. Ideally we throw a JsonException much further down the stack since the converter knows whether or not the value can be null... However I think it's fine to add the checks here for now.

Reply:
As per offline discussion adding `[return: MaybeNull]` attributes, but not adding null check as it better to be checked down stack which will be tracked by separate issue  #1037, added some test cases with active issue number